File: packages/php/src/CapnWeb.php
<?php

declare(strict_types=1);

namespace CapnWeb;

use Closure;
use JsonSerializable;
use Stringable;

/**
 * Error types for RPC exceptions
 */
enum ErrorType: string
{
    case NotFound = 'not_found';
    case Unauthorized = 'unauthorized';
    case Forbidden = 'forbidden';
    case ValidationError = 'validation_error';
    case RateLimited = 'rate_limited';
    case Timeout = 'timeout';
    case ConnectionLost = 'connection_lost';
    case ServerError = 'server_error';
    case Unknown = 'unknown';
}

/**
 * Promise state for tracking resolution
 */
enum PromiseState: string
{
    case Pending = 'pending';
    case Resolved = 'resolved';
    case Rejected = 'rejected';
}

/**
 * RPC Target attribute for exposing objects to server
 */
#[\Attribute(\Attribute::TARGET_CLASS)]
readonly class RpcTarget
{
    public function __construct(
        public ?string $name = null,
    ) {}
}

/**
 * RPC Method attribute for exposing methods
 */
#[\Attribute(\Attribute::TARGET_METHOD)]
readonly class RpcMethod
{
    public function __construct(
        public ?string $name = null,
        public bool $async = false,
    ) {}
}

/**
 * RPC Property attribute for exposing properties
 */
#[\Attribute(\Attribute::TARGET_PROPERTY)]
readonly class RpcProperty
{
    public function __construct(
        public bool $readable = true,
        public bool $writable = false,
    ) {}
}

/**
 * Capture attribute for map expressions
 */
#[\Attribute(\Attribute::TARGET_PARAMETER)]
readonly class Capture
{
    public function __construct(
        public string $name,
    ) {}
}

/**
 * Exception for RPC errors
 */
class RpcException extends \Exception
{
    public function __construct(
        string $message,
        public readonly ErrorType $type = ErrorType::Unknown,
        public readonly ?string $path = null,
        public readonly ?int $retryAfter = null,
        public readonly array $violations = [],
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, 0, $previous);
    }
}

/**
 * Connection exception for network failures
 */
class ConnectionException extends RpcException
{
    public function __construct(
        string $message,
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, ErrorType::ConnectionLost, previous: $previous);
    }
}

/**
 * Timeout exception
 */
class TimeoutException extends RpcException
{
    public function __construct(
        string $message,
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, ErrorType::Timeout, previous: $previous);
    }
}

/**
 * Validation exception with structured errors
 */
class ValidationException extends RpcException
{
    public function __construct(
        string $message,
        array $violations = [],
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, ErrorType::ValidationError, violations: $violations, previous: $previous);
    }
}

/**
 * Represents a captured closure for server-side map operations
 */
readonly class MapExpression implements JsonSerializable
{
    public function __construct(
        public Closure $callback,
        public array $captures = [],
    ) {}

    public function jsonSerialize(): array
    {
        return [
            'type' => 'map_expression',
            'captures' => array_map(
                fn(mixed $value): mixed => $value instanceof RpcPromise
                    ? $value->toReference()
                    : $value,
                $this->captures,
            ),
        ];
    }
}

/**
 * RPC Promise - represents a pending or resolved RPC result
 *
 * Supports promise pipelining and server-side map operations.
 *
 * @template T
 */
class RpcPromise implements Stringable, JsonSerializable
{
    private PromiseState $state = PromiseState::Pending;
    private mixed $value = null;
    private ?RpcException $error = null;

    /** @var array<array{Closure, Closure|null}> */
    private array $callbacks = [];

    /** @var array<string, mixed> */
    private array $pipelinedAccesses = [];

    public function __construct(
        private readonly ?Session $session = null,
        private readonly ?string $callId = null,
        private readonly ?string $method = null,
        private readonly array $args = [],
        private readonly ?RpcPromise $parent = null,
        private readonly ?string $accessPath = null,
    ) {}

    /**
     * Map a function over this promise's result (server-side execution)
     *
     * The callback is serialized and sent to the server, allowing
     * the transform to execute server-side without additional round trips.
     *
     * @template U
     * @param Closure(T): U $callback
     * @return RpcPromise<U>
     */
    public function map(Closure $callback): RpcPromise
    {
        $expression = new MapExpression(
            callback: $callback,
            captures: $this->extractCaptures($callback),
        );

        return new RpcPromise(
            session: $this->session,
            callId: $this->generateCallId(),
            method: '__map__',
            args: [$this, $expression],
            parent: $this,
        );
    }

    /**
     * Alias for map() - remap for server-side collection transformation
     *
     * @template U
     * @param Closure(T): U $callback
     * @return RpcPromise<U>
     */
    public function remap(Closure $callback): RpcPromise
    {
        return $this->map($callback);
    }

    /**
     * Filter the collection server-side
     *
     * @param Closure(T): bool $predicate
     * @return RpcPromise<array<T>>
     */
    public function filter(Closure $predicate): RpcPromise
    {
        $expression = new MapExpression(
            callback: $predicate,
            captures: $this->extractCaptures($predicate),
        );

        return new RpcPromise(
            session: $this->session,
            callId: $this->generateCallId(),
            method: '__filter__',
            args: [$this, $expression],
            parent: $this,
        );
    }

    /**
     * Chain a callback when promise resolves
     *
     * @template U
     * @param Closure(T): U $onResolved
     * @param Closure(RpcException): U|null $onRejected
     * @return RpcPromise<U>
     */
    public function then(Closure $onResolved, ?Closure $onRejected = null): RpcPromise
    {
        $newPromise = new RpcPromise(session: $this->session);

        if ($this->state === PromiseState::Resolved) {
            try {
                $result = $onResolved($this->value);
                $newPromise->resolve($result);
            } catch (\Throwable $e) {
                $newPromise->reject(new RpcException($e->getMessage(), previous: $e));
            }
        } elseif ($this->state === PromiseState::Rejected) {
            if ($onRejected !== null) {
                try {
                    $result = $onRejected($this->error);
                    $newPromise->resolve($result);
                } catch (\Throwable $e) {
                    $newPromise->reject(new RpcException($e->getMessage(), previous: $e));
                }
            } else {
                $newPromise->reject($this->error);
            }
        } else {
            $this->callbacks[] = [
                fn($value) => $newPromise->resolve($onResolved($value)),
                $onRejected !== null
                    ? fn($error) => $newPromise->resolve($onRejected($error))
                    : fn($error) => $newPromise->reject($error),
            ];
        }

        return $newPromise;
    }

    /**
     * Catch rejections
     *
     * @template U
     * @param Closure(RpcException): U $onRejected
     * @return RpcPromise<T|U>
     */
    public function catch(Closure $onRejected): RpcPromise
    {
        return $this->then(fn($v) => $v, $onRejected);
    }

    /**
     * Always execute regardless of outcome
     *
     * @param Closure(): void $callback
     * @return RpcPromise<T>
     */
    public function finally(Closure $callback): RpcPromise
    {
        return $this->then(
            function ($value) use ($callback) {
                $callback();
                return $value;
            },
            function ($error) use ($callback) {
                $callback();
                throw $error;
            },
        );
    }

    /**
     * Wait for the promise to resolve (blocks with Fibers)
     *
     * @return T
     * @throws RpcException
     */
    public function await(): mixed
    {
        if ($this->state === PromiseState::Resolved) {
            return $this->value;
        }

        if ($this->state === PromiseState::Rejected) {
            throw $this->error;
        }

        // In a real implementation, this would use Fibers to suspend
        // and resume when the result arrives. For testing, we execute
        // the request synchronously.
        $this->session?->executeCall($this);

        if ($this->state === PromiseState::Rejected) {
            throw $this->error;
        }

        return $this->value;
    }

    /**
     * Resolve the promise with a value
     *
     * @param T $value
     */
    public function resolve(mixed $value): void
    {
        if ($this->state !== PromiseState::Pending) {
            return;
        }

        $this->state = PromiseState::Resolved;
        $this->value = $value;

        foreach ($this->callbacks as [$onResolved, $_]) {
            $onResolved($value);
        }

        $this->callbacks = [];
    }

    /**
     * Reject the promise with an error
     */
    public function reject(RpcException $error): void
    {
        if ($this->state !== PromiseState::Pending) {
            return;
        }

        $this->state = PromiseState::Rejected;
        $this->error = $error;

        foreach ($this->callbacks as [$_, $onRejected]) {
            if ($onRejected !== null) {
                $onRejected($error);
            }
        }

        $this->callbacks = [];
    }

    /**
     * Check if promise is pending
     */
    public function isPending(): bool
    {
        return $this->state === PromiseState::Pending;
    }

    /**
     * Check if promise is resolved
     */
    public function isResolved(): bool
    {
        return $this->state === PromiseState::Resolved;
    }

    /**
     * Check if promise is rejected
     */
    public function isRejected(): bool
    {
        return $this->state === PromiseState::Rejected;
    }

    /**
     * Get the call ID for this promise
     */
    public function getCallId(): ?string
    {
        return $this->callId;
    }

    /**
     * Get the method name
     */
    public function getMethod(): ?string
    {
        return $this->method;
    }

    /**
     * Get the arguments
     */
    public function getArgs(): array
    {
        return $this->args;
    }

    /**
     * Get the parent promise (for pipelining)
     */
    public function getParent(): ?RpcPromise
    {
        return $this->parent;
    }

    /**
     * Get the access path (for pipelining)
     */
    public function getAccessPath(): ?string
    {
        return $this->accessPath;
    }

    /**
     * Convert to a reference for serialization
     */
    public function toReference(): array
    {
        return [
            'type' => 'promise_ref',
            'callId' => $this->callId,
            'method' => $this->method,
            'args' => $this->args,
        ];
    }

    /**
     * Magic getter for property access (pipelining)
     */
    public function __get(string $name): RpcPromise
    {
        $path = $this->accessPath !== null
            ? "{$this->accessPath}.{$name}"
            : $name;

        if (!isset($this->pipelinedAccesses[$path])) {
            $this->pipelinedAccesses[$path] = new RpcPromise(
                session: $this->session,
                callId: $this->generateCallId(),
                method: '__get__',
                args: [$name],
                parent: $this,
                accessPath: $path,
            );
        }

        return $this->pipelinedAccesses[$path];
    }

    /**
     * Magic method call (pipelining)
     */
    public function __call(string $name, array $arguments): RpcPromise
    {
        return new RpcPromise(
            session: $this->session,
            callId: $this->generateCallId(),
            method: $name,
            args: $arguments,
            parent: $this,
        );
    }

    public function __toString(): string
    {
        return match ($this->state) {
            PromiseState::Pending => '[RpcPromise:pending]',
            PromiseState::Resolved => '[RpcPromise:resolved=' . json_encode($this->value) . ']',
            PromiseState::Rejected => '[RpcPromise:rejected=' . $this->error->getMessage() . ']',
        };
    }

    public function jsonSerialize(): mixed
    {
        return $this->toReference();
    }

    private function generateCallId(): string
    {
        return bin2hex(random_bytes(8));
    }

    /**
     * Extract captured variables from a closure
     */
    private function extractCaptures(Closure $callback): array
    {
        $reflection = new \ReflectionFunction($callback);
        $captures = [];

        foreach ($reflection->getStaticVariables() as $name => $value) {
            $captures[$name] = $value;
        }

        return $captures;
    }
}

/**
 * RPC Session - manages connection to a Cap'n Web server
 */
class Session
{
    /** @var array<string, RpcPromise> */
    private array $pendingCalls = [];

    private bool $connected = false;

    private function __construct(
        private readonly string $url,
        private readonly int $timeout,
        private readonly bool $reconnect,
        private readonly array $headers,
    ) {}

    /**
     * Connect to a Cap'n Web server via WebSocket
     */
    public static function connect(
        string $url,
        int $timeout = 30,
        bool $reconnect = false,
        array $headers = [],
    ): self {
        $session = new self(
            url: $url,
            timeout: $timeout,
            reconnect: $reconnect,
            headers: $headers,
        );

        $session->doConnect();

        return $session;
    }

    /**
     * Create a batch session for HTTP transport
     */
    public static function batch(
        string $url,
        int $timeout = 30,
        array $headers = [],
    ): self {
        return new self(
            url: $url,
            timeout: $timeout,
            reconnect: false,
            headers: $headers,
        );
    }

    /**
     * Execute within a scoped connection
     *
     * @template T
     * @param string $url
     * @param Closure(Session): T $callback
     * @return T
     */
    public static function scoped(string $url, Closure $callback): mixed
    {
        $session = self::connect($url);
        try {
            return $callback($session);
        } finally {
            $session->close();
        }
    }

    /**
     * Magic getter for API access
     */
    public function __get(string $name): RpcPromise
    {
        return new RpcPromise(
            session: $this,
            callId: $this->generateCallId(),
            method: '__get__',
            args: [$name],
        );
    }

    /**
     * Magic method call
     */
    public function __call(string $name, array $arguments): RpcPromise
    {
        return new RpcPromise(
            session: $this,
            callId: $this->generateCallId(),
            method: $name,
            args: $arguments,
        );
    }

    /**
     * Execute multiple calls in parallel
     *
     * @param RpcPromise ...$promises
     * @return array
     */
    public function all(RpcPromise ...$promises): array
    {
        // In a real implementation, this would batch all calls
        // and resolve them in parallel
        return array_map(fn($p) => $p->await(), $promises);
    }

    /**
     * Race multiple promises - first to resolve wins
     *
     * @param RpcPromise ...$promises
     * @return mixed
     */
    public function race(RpcPromise ...$promises): mixed
    {
        // In a real implementation, this would race the requests
        foreach ($promises as $promise) {
            try {
                return $promise->await();
            } catch (RpcException) {
                continue;
            }
        }

        throw new RpcException('All promises rejected');
    }

    /**
     * Execute an RPC call (internal)
     */
    public function executeCall(RpcPromise $promise): void
    {
        $method = $promise->getMethod();
        $args = $promise->getArgs();
        $callId = $promise->getCallId();

        // Build the full call chain for pipelining
        $callChain = $this->buildCallChain($promise);

        try {
            $result = $this->sendRequest($callChain);
            $promise->resolve($result);
        } catch (RpcException $e) {
            $promise->reject($e);
        }
    }

    /**
     * Process incoming messages (for WebSocket)
     */
    public function tick(): void
    {
        // In a real implementation, this would process
        // incoming WebSocket messages
    }

    /**
     * Close the connection
     */
    public function close(): void
    {
        $this->connected = false;
        $this->pendingCalls = [];
    }

    /**
     * Reconnect to the server
     */
    public function reconnect(): void
    {
        $this->close();
        $this->doConnect();
    }

    /**
     * Check if connected
     */
    public function isConnected(): bool
    {
        return $this->connected;
    }

    /**
     * Get the server URL
     */
    public function getUrl(): string
    {
        return $this->url;
    }

    private function doConnect(): void
    {
        // In a real implementation, this would establish
        // the WebSocket connection
        $this->connected = true;
    }

    private function generateCallId(): string
    {
        return bin2hex(random_bytes(8));
    }

    /**
     * Build the full call chain for pipelining
     */
    private function buildCallChain(RpcPromise $promise): array
    {
        $chain = [];
        $current = $promise;

        while ($current !== null) {
            array_unshift($chain, [
                'callId' => $current->getCallId(),
                'method' => $current->getMethod(),
                'args' => $current->getArgs(),
            ]);
            $current = $current->getParent();
        }

        return $chain;
    }

    /**
     * Send an RPC request to the server
     */
    private function sendRequest(array $callChain): mixed
    {
        // This is a stub implementation for testing.
        // In a real implementation, this would:
        // 1. Serialize the call chain to JSON
        // 2. Send over WebSocket or HTTP
        // 3. Parse the response
        // 4. Handle errors

        throw new RpcException(
            'Session not connected to a real server',
            ErrorType::ConnectionLost,
        );
    }
}

/**
 * Helper function to create a duration in seconds
 */
function seconds(int $seconds): int
{
    return $seconds;
}

/**
 * Helper function to create a duration in milliseconds
 */
function milliseconds(int $ms): int
{
    return $ms;
}

/**
 * Await a promise (convenience function)
 *
 * @template T
 * @param RpcPromise<T> $promise
 * @return T
 */
function await(RpcPromise $promise): mixed
{
    return $promise->await();
}

File: packages/dotnet/src/CapnWeb.cs
// ============================================================================
// Cap'n Web SDK for .NET
// Zero-cost capability-based RPC with LINQ-style pipelining
// ============================================================================
//
// Usage:
//   var session = await CapnWeb.Connect("wss://api.example.com");
//   var user = await session.Api.Users().Get(42).Profile().Name;
//
// With LINQ Select() for server-side mapping (eliminates N+1):
//   var squares = await session.Api.GenerateFibonacci(10)
//       .Select(x => session.Api.Square(x));
//
// ============================================================================

using System.Net.WebSockets;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace CapnWeb;

// ============================================================================
// Error Types
// ============================================================================

/// <summary>
/// Base exception for all Cap'n Web RPC errors.
/// </summary>
public class RpcException : Exception
{
    public string ErrorType { get; }
    public int? ErrorCode { get; }

    public RpcException(string message, string errorType = "Error", int? errorCode = null)
        : base(message)
    {
        ErrorType = errorType;
        ErrorCode = errorCode;
    }
}

/// <summary>
/// Thrown when a resource is not found.
/// </summary>
public class NotFoundException : RpcException
{
    public NotFoundException(string message) : base(message, "NotFound", 404) { }
}

/// <summary>
/// Thrown when authentication fails.
/// </summary>
public class UnauthorizedException : RpcException
{
    public UnauthorizedException(string message = "Unauthorized") : base(message, "Unauthorized", 401) { }
}

/// <summary>
/// Thrown when access is forbidden.
/// </summary>
public class ForbiddenException : RpcException
{
    public ForbiddenException(string message) : base(message, "Forbidden", 403) { }
}

/// <summary>
/// Thrown when an argument is invalid.
/// </summary>
public class InvalidArgumentException : RpcException
{
    public InvalidArgumentException(string message) : base(message, "InvalidArgument", 400) { }
}

/// <summary>
/// Thrown when a request times out.
/// </summary>
public class TimeoutException : RpcException
{
    public TimeoutException(string message = "Request timed out") : base(message, "Timeout", 408) { }
}

/// <summary>
/// Thrown when the SDK operation is not yet implemented.
/// </summary>
public class NotImplementedException : RpcException
{
    public NotImplementedException(string message) : base(message, "NotImplemented") { }
}

// ============================================================================
// RPC Promise - The core abstraction for lazy, pipelined RPC calls
// ============================================================================

/// <summary>
/// A lazy RPC result that supports pipelining and LINQ-style operations.
///
/// <para>
/// <c>RpcPromise&lt;T&gt;</c> allows building complex RPC pipelines without
/// immediately executing them. The pipeline is sent to the server in a
/// single round-trip when awaited.
/// </para>
///
/// <example>
/// <code>
/// // Build pipeline - no network calls yet
/// var promise = api.Users().Get(42).Profile().Name;
///
/// // Execute pipeline - one round trip
/// string name = await promise;
/// </code>
/// </example>
/// </summary>
/// <typeparam name="T">The expected result type</typeparam>
public class RpcPromise<T>
{
    private readonly Session _session;
    private readonly List<PipelineStep> _pipeline;
    private readonly MapExpression? _mapExpression;

    internal RpcPromise(Session session, List<PipelineStep> pipeline, MapExpression? mapExpression = null)
    {
        _session = session;
        _pipeline = pipeline;
        _mapExpression = mapExpression;
    }

    internal RpcPromise(Session session, string method, params object?[] args)
        : this(session, [new PipelineStep(method, args.ToList())])
    {
    }

    /// <summary>
    /// Gets the pipeline expression as a string (for debugging/testing).
    /// </summary>
    public string ExpressionString
    {
        get
        {
            var steps = _pipeline.Select(step =>
            {
                var argsStr = string.Join(", ", step.Args.Select(FormatArg));
                return $"{step.Method}({argsStr})";
            });
            var expr = string.Join(".", steps);
            if (_mapExpression != null)
            {
                expr += $".Select({_mapExpression.Expression})";
            }
            return expr;
        }
    }

    private static string FormatArg(object? arg) => arg switch
    {
        null => "null",
        string s => $"\"{s}\"",
        _ => arg.ToString() ?? "null"
    };

    // ========================================================================
    // LINQ-style Select() for server-side mapping (KEY FEATURE!)
    // ========================================================================

    /// <summary>
    /// Maps each element using a server-side transformation.
    ///
    /// <para>
    /// This is the KEY feature of Cap'n Web - it executes the mapping on the
    /// server in a single round-trip, eliminating N+1 query problems.
    /// </para>
    ///
    /// <example>
    /// <code>
    /// // C# LINQ style - single round trip!
    /// var squares = await api.GenerateFibonacci(10)
    ///     .Select(x => api.Square(x));
    /// </code>
    /// </example>
    /// </summary>
    /// <typeparam name="TResult">The result type of the mapping</typeparam>
    /// <param name="selector">The mapping function (executed server-side)</param>
    /// <returns>A new promise for the mapped result</returns>
    public RpcPromise<TResult[]> Select<TResult>(Func<RpcPromise<T>, RpcPromise<TResult>> selector)
    {
        // Create a placeholder promise representing 'x' in the lambda
        var placeholder = new RpcPromise<T>(_session, [new PipelineStep("$element", [])]);

        // Invoke the selector to capture the expression tree
        var resultPromise = selector(placeholder);

        // Extract the expression from the result promise
        var mapExpr = new MapExpression(
            expression: $"x => {resultPromise.ExpressionString}",
            innerPipeline: resultPromise._pipeline.Skip(1).ToList(), // Skip the $element placeholder
            captures: ["$self"]
        );

        return new RpcPromise<TResult[]>(_session, _pipeline, mapExpr);
    }

    /// <summary>
    /// Maps each element using a simple transformation function.
    /// </summary>
    public RpcPromise<TResult[]> Select<TResult>(Func<T, TResult> selector)
    {
        // For simple lambda expressions, we serialize them as expressions
        var mapExpr = new MapExpression(
            expression: "x => transform(x)",
            innerPipeline: [],
            captures: []
        );
        return new RpcPromise<TResult[]>(_session, _pipeline, mapExpr);
    }

    // ========================================================================
    // Pipelining - extend the call chain
    // ========================================================================

    /// <summary>
    /// Extends the pipeline with another method call.
    /// </summary>
    public RpcPromise<TResult> Call<TResult>(string method, params object?[] args)
    {
        var newPipeline = new List<PipelineStep>(_pipeline)
        {
            new(method, args.ToList())
        };
        return new RpcPromise<TResult>(_session, newPipeline);
    }

    /// <summary>
    /// Access a property on the result (creates a pipeline step).
    /// </summary>
    public RpcPromise<TResult> Property<TResult>(string propertyName)
    {
        return Call<TResult>(propertyName);
    }

    // ========================================================================
    // Awaiting - execute the pipeline
    // ========================================================================

    /// <summary>
    /// Gets an awaiter for async/await support.
    /// </summary>
    public TaskAwaiter<T> GetAwaiter() => ExecuteAsync().GetAwaiter();

    /// <summary>
    /// Executes the pipeline and returns the result.
    /// </summary>
    public async Task<T> ExecuteAsync(CancellationToken cancellationToken = default)
    {
        // Stub implementation - returns NotImplemented
        // Real implementation would send pipeline to server via WebSocket
        throw new NotImplementedException(
            $"Pipeline execution not yet implemented. Expression: {ExpressionString}"
        );
    }

    /// <summary>
    /// Implicit conversion to Task for await support.
    /// </summary>
    public static implicit operator Task<T>(RpcPromise<T> promise) => promise.ExecuteAsync();
}

/// <summary>
/// A step in the RPC pipeline.
/// </summary>
/// <param name="Method">The method name to call</param>
/// <param name="Args">Arguments to pass to the method</param>
internal record PipelineStep(string Method, List<object?> Args);

/// <summary>
/// Represents a map/remap expression for server-side collection transformation.
/// </summary>
/// <param name="Expression">The lambda expression string</param>
/// <param name="InnerPipeline">The pipeline steps inside the map</param>
/// <param name="Captures">Variables captured from outer scope (e.g., $self)</param>
internal record MapExpression(string Expression, List<PipelineStep> InnerPipeline, List<string> Captures);

// ============================================================================
// Session - Connection to a Cap'n Web server
// ============================================================================

/// <summary>
/// A session connected to a Cap'n Web server.
/// </summary>
public class Session : IAsyncDisposable
{
    private readonly string _url;
    private ClientWebSocket? _webSocket;
    private bool _isConnected;

    internal Session(string url)
    {
        _url = url;
        _isConnected = false;
    }

    /// <summary>
    /// Gets the URL of this session.
    /// </summary>
    public string Url => _url;

    /// <summary>
    /// Gets whether the session is connected.
    /// </summary>
    public bool IsConnected => _isConnected;

    /// <summary>
    /// Gets the dynamic API for making RPC calls.
    /// </summary>
    public DynamicApi Api => new(this);

    /// <summary>
    /// Connects to the server.
    /// </summary>
    internal async Task ConnectAsync(CancellationToken cancellationToken = default)
    {
        _webSocket = new ClientWebSocket();

        // Convert HTTP URL to WebSocket URL
        var wsUrl = _url
            .Replace("http://", "ws://")
            .Replace("https://", "wss://");

        try
        {
            await _webSocket.ConnectAsync(new Uri(wsUrl), cancellationToken);
            _isConnected = true;
        }
        catch (Exception ex)
        {
            throw new RpcException($"Failed to connect to {wsUrl}: {ex.Message}", "ConnectionFailed");
        }
    }

    /// <summary>
    /// Closes the session.
    /// </summary>
    public async Task CloseAsync(CancellationToken cancellationToken = default)
    {
        if (_webSocket is { State: WebSocketState.Open })
        {
            await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Session closed", cancellationToken);
        }
        _isConnected = false;
    }

    /// <summary>
    /// Disposes the session.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        await CloseAsync();
        _webSocket?.Dispose();
        GC.SuppressFinalize(this);
    }

    // ========================================================================
    // Dynamic method invocation (for conformance testing)
    // ========================================================================

    /// <summary>
    /// Calls a method dynamically (for conformance testing).
    /// </summary>
    public async Task<JsonNode?> CallMethodAsync(string method, List<JsonNode?> args, CancellationToken cancellationToken = default)
    {
        // Stub implementation
        throw new NotImplementedException($"Method call not implemented: {method}({string.Join(", ", args)})");
    }

    /// <summary>
    /// Executes a pipeline dynamically (for conformance testing).
    /// </summary>
    public async Task<JsonNode?> CallPipelineAsync(List<DynamicPipelineStep> steps, CancellationToken cancellationToken = default)
    {
        // Stub implementation
        var expr = string.Join(".", steps.Select(s => $"{s.Call}({string.Join(", ", s.Args)})"));
        throw new NotImplementedException($"Pipeline not implemented: {expr}");
    }

    /// <summary>
    /// Calls a method with a map expression (for conformance testing).
    /// </summary>
    public async Task<JsonNode?> CallWithMapAsync(
        string method,
        List<JsonNode?> args,
        DynamicMapExpression? map,
        CancellationToken cancellationToken = default)
    {
        // Stub implementation
        var expr = $"{method}({string.Join(", ", args)})";
        if (map != null)
        {
            expr += $".Select({map.Expression})";
        }
        throw new NotImplementedException($"Map call not implemented: {expr}");
    }
}

/// <summary>
/// A dynamic pipeline step for conformance testing.
/// </summary>
/// <param name="Call">The method to call</param>
/// <param name="Args">Arguments to pass</param>
/// <param name="Alias">Optional alias for the result</param>
public record DynamicPipelineStep(string Call, List<JsonNode?> Args, string? Alias = null);

/// <summary>
/// A dynamic map expression for conformance testing.
/// </summary>
/// <param name="Expression">The lambda expression string</param>
/// <param name="Captures">Variables captured from outer scope</param>
public record DynamicMapExpression(string Expression, List<string> Captures);

// ============================================================================
// Dynamic API - Untyped access to RPC methods
// ============================================================================

/// <summary>
/// Dynamic API for making untyped RPC calls.
/// </summary>
public class DynamicApi
{
    private readonly Session _session;

    internal DynamicApi(Session session)
    {
        _session = session;
    }

    /// <summary>
    /// Creates a promise for a method call.
    /// </summary>
    public RpcPromise<T> Call<T>(string method, params object?[] args)
    {
        return new RpcPromise<T>(_session, method, args);
    }

    /// <summary>
    /// Creates a promise for Square(n).
    /// </summary>
    public RpcPromise<int> Square(int n) => Call<int>("square", n);

    /// <summary>
    /// Creates a promise for Square(n) with dynamic argument.
    /// </summary>
    public RpcPromise<int> Square(RpcPromise<int> n) => Call<int>("square", n);

    /// <summary>
    /// Creates a promise for ReturnNumber(n).
    /// </summary>
    public RpcPromise<double> ReturnNumber(double n) => Call<double>("returnNumber", n);

    /// <summary>
    /// Creates a promise for ReturnNull().
    /// </summary>
    public RpcPromise<object?> ReturnNull() => Call<object?>("returnNull");

    /// <summary>
    /// Creates a promise for ReturnUndefined().
    /// </summary>
    public RpcPromise<object?> ReturnUndefined() => Call<object?>("returnUndefined");

    /// <summary>
    /// Creates a promise for GenerateFibonacci(count).
    /// </summary>
    public RpcPromise<int[]> GenerateFibonacci(int count) => Call<int[]>("generateFibonacci", count);

    /// <summary>
    /// Creates a promise for MakeCounter(initial).
    /// </summary>
    public RpcPromise<Counter> MakeCounter(int initial) => Call<Counter>("makeCounter", initial);

    /// <summary>
    /// Creates a promise for ThrowError().
    /// </summary>
    public RpcPromise<object> ThrowError() => Call<object>("throwError");
}

/// <summary>
/// Represents a Counter capability.
/// </summary>
public class Counter
{
    private readonly Session _session;
    private readonly List<PipelineStep> _pipeline;

    internal Counter(Session session, List<PipelineStep> pipeline)
    {
        _session = session;
        _pipeline = pipeline;
    }

    /// <summary>
    /// Gets the current value of the counter.
    /// </summary>
    public RpcPromise<int> Value()
    {
        var newPipeline = new List<PipelineStep>(_pipeline)
        {
            new("value", [])
        };
        return new RpcPromise<int>(_session, newPipeline);
    }

    /// <summary>
    /// Increments the counter by the specified amount.
    /// </summary>
    public RpcPromise<int> Increment(int amount)
    {
        var newPipeline = new List<PipelineStep>(_pipeline)
        {
            new("increment", [amount])
        };
        return new RpcPromise<int>(_session, newPipeline);
    }
}

// ============================================================================
// Static Entry Point
// ============================================================================

/// <summary>
/// Static entry point for connecting to Cap'n Web servers.
/// </summary>
public static class CapnWebClient
{
    /// <summary>
    /// Connects to a Cap'n Web server.
    /// </summary>
    /// <param name="url">The URL to connect to (HTTP or WebSocket)</param>
    /// <param name="cancellationToken">Optional cancellation token</param>
    /// <returns>A connected session</returns>
    /// <example>
    /// <code>
    /// await using var session = await CapnWebClient.Connect("wss://api.example.com");
    /// var result = await session.Api.Square(5);
    /// </code>
    /// </example>
    public static async Task<Session> Connect(string url, CancellationToken cancellationToken = default)
    {
        var session = new Session(url);
        await session.ConnectAsync(cancellationToken);
        return session;
    }

    /// <summary>
    /// Gets whether the SDK is fully implemented.
    /// </summary>
    public static bool IsImplemented => false;

    /// <summary>
    /// Gets the SDK version.
    /// </summary>
    public static string Version => "0.1.0";
}
